rust/src/checkpoints.rs

```diff
@@ -0,0 +1,196 @@
+
+        Self {
+            table_uri: table_uri.to_string(),
+            delta_log_uri,
+            last_checkpoint_uri,
+            storage,
+        }
+    }
+
+    /// Creates a new checkpoint at the specified version.
+    /// NOTE: This method loads a new instance of delta table to determine the state to
+    /// checkpoint.
+    pub async fn create_checkpoint_for_version(
+        &self,
+        version: DeltaDataTypeVersion,
+    ) -> Result<(), CheckPointWriterError> {
+        let table = open_table_with_version(self.table_uri.as_str(), version).await?;
+
+        self.create_checkpoint_from_state(version, table.get_state())
+            .await
+    }
+
+    /// Creates a new checkpoint at the specified version from the given DeltaTableState.
+    pub async fn create_checkpoint_from_state(
+        &self,
+        version: DeltaDataTypeVersion,
+        state: &DeltaTableState,
+    ) -> Result<(), CheckPointWriterError> {
+        // TODO: checkpoints _can_ be multi-part... haven't actually found a good reference for
+        // an appropriate split point yet though so only writing a single part currently.
+        // See https://github.com/delta-io/delta-rs/issues/288
+
+        info!("Writing parquet bytes to checkpoint buffer.");
+        let parquet_bytes = self.parquet_bytes_from_state(state)?;
+
+        let size = parquet_bytes.len() as i64;
+
+        let checkpoint = CheckPoint::new(version, size, None);
+
+        let file_name = format!("{:020}.checkpoint.parquet", version);
+        let checkpoint_uri = self.storage.join_path(&self.delta_log_uri, &file_name);
+
+        info!("Writing checkpoint to {:?}.", checkpoint_uri);
+        self.storage
+            .put_obj(&checkpoint_uri, &parquet_bytes)
+            .await?;
+
+        let last_checkpoint_content: serde_json::Value = serde_json::to_value(&checkpoint)?;
+        let last_checkpoint_content = serde_json::to_string(&last_checkpoint_content)?;
+
+        info!(
+            "Writing _last_checkpoint to {:?}.",
+            self.last_checkpoint_uri
+        );
+        self.storage
+            .put_obj(
+                self.last_checkpoint_uri.as_str(),
+                last_checkpoint_content.as_bytes(),
+            )
+            .await?;
+
+        Ok(())
+    }
+
+    fn parquet_bytes_from_state(
+        &self,
+        state: &DeltaTableState,
+    ) -> Result<Vec<u8>, CheckPointWriterError> {
+        let current_metadata = state
+            .current_metadata()
+            .ok_or(CheckPointWriterError::MissingMetaData)?;
+        let jsons: Vec<Result<serde_json::Value, ArrowError>> = vec![
+            action::Action::protocol(action::Protocol {
+                min_reader_version: state.min_reader_version(),
+                min_writer_version: state.min_writer_version(),
+            }),
+            action::Action::metaData(action::MetaData::try_from(current_metadata.clone())?),
+        ]
+        .into_iter()
```

Comment:
I was able to get `once` working using the following code:

```diff
-        let jsons: Vec<Result<serde_json::Value, ArrowError>> = vec![
-            action::Action::protocol(action::Protocol {
-                min_reader_version: state.min_reader_version(),
-                min_writer_version: state.min_writer_version(),
-            }),
-            action::Action::metaData(action::MetaData::try_from(current_metadata.clone())?),
-        ]
-        .into_iter()
+        let jsons = std::iter::once(action::Action::protocol(action::Protocol {
+            min_reader_version: state.min_reader_version(),
+            min_writer_version: state.min_writer_version(),
+        }))
+        .chain(std::iter::once(action::Action::metaData(
+            action::MetaData::try_from(current_metadata.clone())?,
+        )))
```

Reply:
fixed. super cool - i learned a lot about iterators thanks to this series of comments. :bow: